share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java

```diff
@@ -537,38 +539,322 @@ private static ShareGroupOffset merge(ShareGroupOffset soFar, ShareUpdateValue n
+            // as a prefix batch which is non overlapping will only
+            // be checked once.
+            if (overlapState.nonOverlapping() != null) {
+                overlapState.nonOverlapping().forEach(sortedBatches::remove);
+                finalBatches.addAll(overlapState.nonOverlapping());
+            }
+
+            if (candidate == null) {
+                overlapState = BatchOverlapState.SENTINEL;
+                continue;
+            }
+
+            // overlap and same state (last.firstOffset <= candidate.firstOffset) due to sort
+            // covers:
+            // case:        1        2          3            4          5           6          7 (contiguous)
+            // last:        ______   _______    _______      _______   _______   ________    _______
+            // candidate:   ______   ____       __________     ___        ____       _______        _______
+            if (compareBatchState(candidate, last) == 0) {
+                sortedBatches.remove(last);  // remove older smaller interval
+                sortedBatches.remove(candidate);
+
+                last = new PersisterStateBatch(
+                    last.firstOffset(),
+                    // cover cases
+                    // last:      ______   ________       _________
+                    // candidate:   ___       __________           _____
+                    Math.max(candidate.lastOffset(), last.lastOffset()),
+                    last.deliveryState(),
+                    last.deliveryCount()
+                );
+
+                sortedBatches.add(last);
+            } else if (candidate.firstOffset() <= last.lastOffset()) { // non-contiguous overlap
+                // overlap and different state
+                // covers:
+                // case:        1        2 (NP)      3            4          5           6
+                // last:        ______   _______    _______      _______    _______     ________
+                // candidate:   ______   ____       _________      ____        ____          _______
+                // max batches: 1           2       2                3          2            2
+                // min batches: 1           1       1                1          1            2
+
+                // remove both last and candidate for easier
+                // assessment about adding batches to sortedBatches
+                sortedBatches.remove(last);
+                sortedBatches.remove(candidate);
+
+                if (candidate.firstOffset() == last.firstOffset()) {
+                    if (candidate.lastOffset() == last.lastOffset()) {  // case 1
+                        // candidate can never have lower or equal priority
+                        // since sortedBatches order takes that into account.
+                        sortedBatches.add(candidate);
+                    } else {
+                        // case 2 is not possible with TreeSet. It is symmetric to case 3.
```

Comment:
So I suppose that case 2 would actually be case 3 because of the sorting order of the key.

Reply:
Yes that true, have kept the example for completeness sake